### DISCUSSION TITLE
CLI/NPM extensions

### DISCUSSION THREAD

**Discussion by WoLfulus** [directus/directus#5117] [Feature Requests]
# Extensions

We've been talking about supporting npm extension for a while now, and while thinking/working/designing the CLI over the past few weeks, there are some inputs I'd like to give to help shaping our extension requirements.

The idea here is to discuss if we can leverage some of the stuff for the npm extensions too (like the discovery/loading engine for example), and to raise if there are additional use cases that I didn't think about.

So, here's a quick overview of what I found out to be nice to have and how I imagine it working. 

## Notes before we start

All of the following sections are based on some premises regarding the CLI, and how it works.

Also, when mentioning these, I'll be referring to the ones in front.

| name | package |
| ---- | -------- |
| `API` | `directus` |
| `SDK` | `@directus/sdk` |
| `CLI` | `@directus/cli` |

### 1. Different packages

One of the changes I'm making in the CLI is having a package called `@directus/cli` to handl CLI stuff. This package contains the CLI logic and goodies to work with CLI commands, I/O, configs and params.

The API package can also contain a `directus` command in case user doesn't install the CLI package (or wants to use `npx directus x` instead). This is only an alias that will execute `@directus/cli/bin/directus` instead.

### 2. CLI doesn't know the API (by default)

The package only exposes commands for what we have in the SDK itself, and it works by talking to a Directus instance through HTTP(s), thus, only having access to what the authenticated user already have.

Commands would do exactly what the SDK has (aiming for 100% feature coverage), plus some commands to manage which instance we're connected to:

- `directus instance connect dev --url http://localhost:8055`
- `directus instance connect prod --url https://prod.mydomain.com`
- `directus instance disconnect dev`
- `directus instance list`
- `directus instance use dev`
- `directus some command --instance=prod`

### 3. CLI knows the API (indirectly)

When a command is executed within a project folder (with dependencies installed), the CLI will be able to find and load external commands from the api package (`directus`) as if it was an extension (keep this in mind). After loading it, it will expose a group of commands to the CLI automatically (dictated by the API itself).

Example:

- `directus server bootstrap`
- `directus server start`

### 4. CLI handles configs

CLI will load configs and merge them when necessary, making them available to the commands upon execution.

## 5. CLI raises events

CLI has an event system similar to what the API has. It's used by extensions to register custom behavior and add/change CLI output when needed. These can also be used to add more features to the CLI, like supporting a different output format (XML for example).

### 6. CLI has all the information you need

All CLI commands should have proper (and full) documentation about themselves. User should be able to quickly read, check usage, synopsis and look at some examples just by adding a `--help` flag to the command.

#### Project level

CLI tries to load configuration files starting from the working directory, going down until it hits the root folder. The files it will search for are:

- `package.json` on `directus` key
- `.directus(rc)?.ya?ml`
- `.directus(rc)?.json`
- `.directus(rc)?.js`

> The first match is the only one that gets loaded.

Examples of what we use these for:

- The default instance name to use (can override with --instance)
- \*Extension names to be loaded (when it doesn't match the pattern)
- Anything that isn't sensitive
- Custom CLI extension configs (CLI)

> \* More about this later

### System level

CLI will load `~/.directus/directus.yml` for system/machine level configs.

Examples of what we use these for:

- Anything that isn't sensitive and can leak when pushed to a repository
- The default (global) instance name to use when using CLI outside a project 
  directory (can override with --instance)
- Instance authentication tokens ("local storage")
- Anything that can be remembered based on the current working directory

Commands can read and update these settings.

## CLI extensions

Now that we know what the CLI can do, this is where it starts to get 
interesting.

We can automatically load these extensions when they are installed in the project folder. This allows commands to be available based on the context the CLI is running. It also allows for reliable automation when in a CI environment. On top of that, commands can also be loaded from the project folder itself, like `./extensions/cli` (project-specific commands).

> NOTE: We can also install them globally inside `~/.directus/node_modules`

The CLI will load all the available commands on that package using a pattern
we define, for example: 

- `directus-*-cli`
- `@directus/cli-*`
 
The loading process can be tweaked by the extension's own `package.json` (inside "directus-extension" key) in case it's impossible to comply with the pattern we defined for some reason.

It's also possible to specify dependencies in the extension's `package.json` which will be used in a topological sort of packages to be loaded.

### Possibilities

This should cover pretty much everything we need to have proper tooling while keeping each CLI scoped to what it does (for both internal and community-driven ones). Here are some examples IMHO makes sense being developed as CLI extensions:

#### Jonesus's extension toolkit

> https://github.com/directus-community/extension-toolkit

- `directus extension path ./extensions`
- `directus extension create display my-display --typescript`
- `directus extension create module my-module --typescript`
- `directus extension create hook my-hook`
- `directus extension compile --all`

Can expand to project level generators

- `directus project create --typescript`
- `directus project compile`

#### Data migration tools

- `directus migration create --instance=dev > dev.yml`
- `directus migration apply --instance=staging --file=dev.yml`

#### Vendor specific interactions

- `directus aws create --ec2=t2.large`
- `directus aws deploy`
- `directus digitalocean create --ec2=t2.large`
- `directus aws deploy`

#### Deployment workflows

- `directus deploy target create do-dev --digitalocean`
- `directus deploy target create pm2-dev --pm2`
- `directus deploy target create aws-prod --aws`
- `directus deploy target list`
- `directus deploy target set do-dev`
- `directus deploy`
- `directus deploy --target=pm2-dev`
- `directus deploy --target=aws-prod`

#### Project specific commands

- `directus myproject billing --customer=42 --from=04/01/2021 --to=04/30/2021 --format=csv > april.csv`
- `directus myproject backup`
- `directus myproject build --deploy`

#### Integrations

This one I already have developed and published to NPM, just needs a conversion to a CLI extension since it's on it's own package right now. This can definitely land as a PoC `next.js` integration already.

> .directusrc.yml

```yml
next:
  port: 3000
  routes:
    next:
      # Serves the Next website through both domains
      - "app.local"
    directus:
      # Serves Directus instead of Next if accessed through "directus.app.local"
      - "directus.app.local"
```

> Starts Directus and Next.js on the same process/express application, manages which one will be served based on route rules.

`directus next start`

**Comment by WoLfulus** [comment]
@rijkvanzanten 

> What do you mean by "could use it as an engine for other extensions"?
> I see the CLI as a stand-alone tool to interact with the api, so I'm failing to see the connection with the "regular" extensions (eg modules/layouts/hooks)
> I get the idea of extending the CLI itself in a similar fashion, just don't really understand what you meant by that

It falls in the server example, the CLI is a stand-alone tool, yes, but given the proper context it can be smart enough to load additional stuff (like knowing it's on a project folder by seeing that `directus` package is installed and showing it's commands too)
The CLI can expose the discovered packages to the the available commands (thus making server able to query another package's metadata, checking if they have modules/displays/etc and making it available in the server itself)

This is really similar to what Laravel does, but with way less stuff involved.

**Reply by benhaynes** [reply]
I read the whole post above, but won't pretend to understand the detail of it all. I'll leave this one to you, Rijk, and the rest of the contributing team! 💯

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
